Replace `flask` `JSONEncoder` by `JSONProvider`

mbercx opened the pull request:
Fixes #5986 

`JSONEncoder` was deprecated in `flask==2.2.0` and removed in `flask==2.3.0`. Here it is replaced by `JSONProvider`.

mbercx commented:
@eimrek maybe you can have a look at this? I'm completely unfamiliar with `flask` I'm afraid. ^^

mbercx commented on pyproject.toml:
```diff
@@ -70,7 +70,7 @@ ssh_kerberos = [
     "flask-cors~=3.0",
     "flask-restful~=0.3.7",
-    "flask~=2.0",
+    "flask>=2.0<3.0",
```
For sure, I clearly wasn't very awake yet to make sure a typo. 😅 Thanks!

mbercx commented on aiida/restapi/common/utils.py:
```diff
@@ -27,43 +27,40 @@
 
 ########################## Classes #####################
-class CustomJSONEncoder(JSONEncoder):
+class CustomJSONProvider(JSONProvider):
```
Fair point! Wasn't aware of this one, just looking for an easy fix.

mbercx commented on aiida/restapi/common/utils.py:
```diff
@@ -27,43 +27,40 @@
 
         # If not returned yet, do it in the default way
-        return JSONEncoder.default(self, o)
+        return json.dumps(self, obj, **kwargs)  # pylint: disable=too-many-function-args
```
I guess I can also just remove it then?

eimrek commented:
Hi @sphuber @mbercx I promised to take a shot at this, will focus on it now

eimrek commented:
@sphuber @mbercx this should be fixed, tests are passing. But i'm not sure why the pre-commit fails for a line that was not touched & not sure how to deal with it.

sphuber commented:
Thanks @eimrek . That seems to be `mypy` complainin, not sure why though, because it isn't explicitly updated. @mbercx is there a reason why all the dependencies in the `requirements.txt` files need to be updated? Can we really limit this PR just to removing pin of flask and fixing the REST API implementation?

mbercx commented:
> is there a reason why all the dependencies in the requirements.txt files need to be updated? Can we really limit this PR just to removing pin of flask and fixing the REST API implementation?

I may have gone a little overboard, but adapting the flask dependency alone definitely wasn't enough. So I used `pip-tools` to "compile" a proper new `requirements.txt` file for each Python version. Seemed like the best approach? 

I think I can "fix" the `mypy` issue, just a sec.

mbercx commented:
![mexican_cat](https://github.com/aiidateam/aiida-core/assets/25607960/0756ace3-51e1-43c6-ab2b-e22bef03216a)

sphuber commented:
> I may have gone a little overboard, but adapting the flask dependency alone definitely wasn't enough. So I used pip-tools to "compile" a proper new requirements.txt file for each Python version. Seemed like the best approach?

I think it would be instructive rather to know the minimal required set of changes to address the problem and not change the requirements completely. These are autogenerated by the `test-install.yml` workflow. And why does this have to add/remove mypy/pylint exclude rules? Why are the pre-commit dependencies updated? These are not installed through the requirements files. I don't understand...

mbercx commented:
> I think it would be instructive rather to know the minimal required set of changes to address the problem and not change the requirements completely. These are autogenerated by the test-install.yml workflow. And why does this have to add/remove mypy/pylint exclude rules? Why are the pre-commit dependencies updated? These are not installed through the requirements files. I don't understand...

Heh, sorry, I didn't mean to make things more complicated. Let me explain how we got here:

1. I wanted to review https://github.com/aiidateam/aiida-core/pull/5985, but saw the docs were failing.
2. I updated the `flask` dependency locally, but then messed up the version in the `pyproject.toml` [as you noted here](https://github.com/aiidateam/aiida-core/pull/5987#discussion_r1180443229). I also then found out we had separate `requirements` files for testing (still not really sure why this is _needed_. Can't we just install from `pyproject.toml` for the tests as well?)
3. I first tried to update the `requirements` manually, but quickly found out that it's tricky to do this and pass the jobs in the `test-install.yaml` workflow. So I thought there must be a better way to do this.
4. I read the "Dependency management" part of the wiki, specifically [_this section_](https://github.com/aiidateam/aiida-core/wiki/AiiDA-Dependency-Management#2-update-the-requirements-files). Realised that I had opened the PR from a fork and that's probably why the automated workflow that updates the `requirements` didn't trigger. So I tried https://github.com/aiidateam/aiida-core/pull/5990, but alas. Maybe I still did something wrong?
5. So I looked into how to "properly" create requirements files that are consistent with a `pyproject.toml`, and found [`pip-tools`](https://pypi.org/project/pip-tools/). I used that for the different Python versions (wanted to look into using `tox` for this, but just wanted to get it done at this point), and voila: that's how I got all the new `requirements` files.
6. Changing all these versions did introduce some `pylint`/`mypy` issues, so I fixed these.

Although I do understand how it's nice to have the "minimal" set of required changes for the `requirements` for clarity, the requirements in this PR should be what you obtain using a `pyproject.toml` install (or at least what you did two weeks ago). I'm not sure how to generate this "minimal" set of changes automatically either, I'm afraid.

On a different note: we should appoint a new dependency manager (DM), now that @csadorf has unfortunately left our team (and is clearly missed! 🙃). Although I have a somewhat affectionate disposition towards this acronym from my DnD days, I think it will be hard to find a volunteer. I've made [a note in the HackMD notes](https://hackmd.io/YCY8ryoESGCZMRWS6Sn2BA?view).

mbercx commented:
> "vierkantemillimeterzeiker"

How can I be upset when you use such beautiful Dutch expressions? 😄 

> was probably just introduced by an accumulation of small incorrect steps.

A fair assumption, it turns out! 

> Note that just changing flask was sufficient, so not quite sure why you ran into trouble and had to update a whole bunch of other dependencies.

Indeed. It was still very early, but I tried doing this [here](https://github.com/aiidateam/aiida-core/pull/5990/files#diff-f909b7f9666d83397d43aed4f45b4ca2cbbf565c8ac6e030853613a67869d325). I did pick `2.3.1` when going down this rabbit hole, not the minimal required `2.2`. Perhaps that caused more issues. 🤷 

> To illustrate it, I created a PR with the minimal changes, with passes the tests just fine: https://github.com/aiidateam/aiida-core/pull/6011

I saw 😅. I will admit for this one I was rooting for the tests to fail for the first time 😉. Closing this PR with this comment.

> The reason for using requirements in the tests is to guarantee consistency. This is especially important for the pre-commit job of the CI that we don't want to fail just because another library released an update. Note that we still actually run the unit tests against an environment that is created from the pyproject.toml by doing pip install -e .[atomic_tools,docs,notebook,rest,tests]. This simulates the situation where a user installs it this way and we make sure the tests still pass.

Yes, I realised it after a conversation with @chrisjsewell this morning. Makes sense when dealing with `aiida-core` and its massive dependency list. 👍 

> I have de-facto been doing this, so I just went ahead and assigned myself in the wiki. Like you, I think there will be few volunteers. But happy to stand aside if someone else is really keen to take it on.

Thank you for your service. 🙏 It's a thankless job, but someone's got to do it.
